HW2/P3/P3.txt

```diff
@@ -0,0 +1,24 @@
+P3_graph compares the performance across 1, 2 and 4 threads, both with and without instruction-level parallelism via AVX.  There is a slight variation in runtime across runtimes, so I have included five different runtimes for each option for completeness.
+
+Without instruction-level parallelism:
+Performance with one thread is roughly similar to the previous version of the code that did not include threading.  A significant speed-up was observed when increasing the number of threads: the 2-threaded code ran approximately 2 times faster than the 1-threaded version; and the 4-threaded code ran slightly less than 4 times faster (about 3.8x).
+
+With instruction-level parallelism:
+Performance with one thread is significantly worse than both the original serial code and the one threaded version without instruction-level parallelism. It appears that the overhead introduced by this approach offsets the speed-up achieved by the 8-way parallelism. However, performance is significantly better with instruction-level parallelism for both 2 threads and 4 threads; in fact, the run-time is over ten times faster with four threads than with one thread. It is interesting to note that the majority of the speed-up is already obtained when using two threads (approx. 8x speed-up).
```

Comment:
I've been wondering that myself while doing my code reviews, as I noticed that most people saw a speed-up with instruction-level parallelism and one thread.  I've compared my code to a few different versions and nothing stands out as being materially different, so would be interested in any ideas of what might be causing this!
(I also just re-ran it, to make sure it wasn't a typo...)